app/sample_manifest_excel/sample_manifest_excel/upload/processor/tube_rack.rb

```diff
@@ -0,0 +1,201 @@
+# frozen_string_literal: true
+
+require 'net/http'
+require 'json'
+require_dependency 'sample_manifest_excel/upload/processor/base'
+
+module SampleManifestExcel
+  module Upload
+    module Processor
+      # TODO: had to explicitly specify the namespace for Base here otherwise it picks up Upload::Base
+      # Used for processing the upload of sample manifests.
+      # Contains behaviour specific to processing 'Tube Rack' manifests.
+      class TubeRack < SampleManifestExcel::Upload::Processor::Base
+        include ActiveModel::Validations
+
+        validates_presence_of :tube_rack_barcodes_from_manifest
+        attr_reader :tube_rack_barcodes_from_manifest
+
+        def initialize(upload)
+          super(upload)
+          @tube_rack_information_processed = false
+          @tube_rack_barcodes_from_manifest = retrieve_tube_rack_barcodes_from_manifest
+          @tube_barcodes_from_manifest = @upload.data.column(1).compact # TODO: do this based on the column name, not number
+          @should_process_tube_rack_information = should_process_tube_rack_information?
+        end
+
+        def run(tag_group)
+          return unless valid?
+
+          if @should_process_tube_rack_information
+            @rack_size = @upload.sample_manifest.tube_rack_purpose.size
+            return unless retrieve_scan_results && validate_against_scan_results && validate_coordinates(@rack_size, @rack_barcode_to_scan_results)
```

Comment:
These validations can happen in the line 28 ("return unless valid?") if you move them as validation callbacks

Reply:
Hey yeah, I wasn't sure what the best way to do this was. I think the issue I had with putting them as validation call backs was that they are important for setting variables that the later method uses (e.g. @rack_barcode_to_scan_results) - so they are actually processing as well as validating, and need to be called sequentially. I don't really like how they are laid out at the moment though...

Reply:
The validations could be ran inside the run process whenever you call valid? or validate, so the validation can happen at any moment you decide, not necessarily before starting processing. Anyway, the validations make more sense when is with a model that is going to be saved in the database, which is not the case.